### PR TITLE
Finding datanodes & new replication schema

### PR DESCRIPTION
**Pull request by prb2**
Addresses #60 

Currently fails ReadWriteTest, will continue looking into it and will add updates to this PR.

### CODE REVIEW

File: zookeeper/include/zk_nn_client.h [hunk 1]
```diff
@@ -28,6 +28,23 @@ typedef struct
...
+{
+	std::string dn_id;
+	uint64_t free_bytes;	//free space on disk
+	uint32_t num_xmits;		//current number of xmits
+
+	TargetDN(std::string id, int bytes, int xmits) : dn_id(id), free_bytes(bytes), num_xmits(xmits)
+	{
+	}
+
+	bool operator<(const struct TargetDN& other) const
```
Comment (Swyrough) [inline zookeeper/include/zk_nn_client.h:41]: IS this minimizing? If tdn1 < tnd2 then tdn1.xmits > tdn2.xmits, thats what we want

File: zookeeper/include/zk_nn_client.h [hunk 2]
```diff
@@ -180,11 +198,11 @@ class ZkNnClient : public ZkClientCommon {
 		const int IS_FILE = 2;
 		const int IS_DIR = 1;
 		// TODO: Should eventually be read from a conf file
-		const int ACK_TIMEOUT = 60; // 60 second timeout when waiting for replication acknowledgements
+		const int ACK_TIMEOUT = 600000; // in millisecons, 10 minute timeout when waiting for replication acknowledgements
+		const int BLOCKSIZE = 64; // TODO: Is this a const or should this be read for each block?
```
Comment (Swyrough) [inline zookeeper/include/zk_nn_client.h:202]: Blocksize is on a per file basis, I dont think it should be a const

File: zookeeper/source/zk_nn_client.cc [hunk 3]
```diff
@@ -21,359 +22,359 @@
...
+
+	bool ZkNnClient::file_exists(const std::string& path) {
+		bool exists;
+		int error_code;
+		if (zk->exists(ZookeeperPath(path), exists, error_code)) {
+			return exists;
+		} else {
+			// TODO: Handle error
+		}
+	}
```
Comment (Swyrough) [inline zookeeper/source/zk_nn_client.cc:83]: should return false here, what does c++ default the return value to?

File: zookeeper/source/zk_nn_client.cc [hunk 4]
```diff
@@ -21,359 +22,359 @@
...
 	}
-	return true;
-    }
 
-    bool ZkNnClient::add_block(AddBlockRequestProto& req, AddBlockResponseProto& res) {
+
+	bool ZkNnClient::previousBlockComplete(uint64_t prev_id) {
+		int error_code;
+		/* this value will eventually be read from config file */
+		int MIN_REPLICATION = 1;
```
Comment (Swyrough) [inline zookeeper/source/zk_nn_client.cc:108]: I believe this is also on a per-file basis. It is a field in a file znode. So we would need to read the file that the block belongs to...is it even possible to get a blockid -> file mapping?

**Review by Swyrough — COMMENTED** [review]
Do you know why there are huge chunks of the file that it says you changed when it just looks like its been moved or something? IT is kind of hard to review the changes when there is so much red/green

File: zookeeper/include/zk_nn_client.h (hunk 1 above)
Comment (prb2) [reply zookeeper/include/zk_nn_client.h:41]: Yep, the fewer the number of transmits, the higher the priority.

File: zookeeper/include/zk_nn_client.h (hunk 2 above)
Comment (prb2) [reply zookeeper/include/zk_nn_client.h:202]: Yep, this is fixed to look up the individual blocksizes.

File: zookeeper/source/zk_nn_client.cc (hunk 3 above)
Comment (prb2) [reply zookeeper/source/zk_nn_client.cc:83]: I'm not really sure, I wasn't the author of this method. The latest code at this location does have a log statement.

File: zookeeper/source/zk_nn_client.cc (hunk 4 above)
Comment (prb2) [reply zookeeper/source/zk_nn_client.cc:108]: I'm not really sure, I wasn't the author of this method.

**Comment by prb2** [comment]
@Swyrough Yea, the reason for the huge chunks of changes (and the reason things I didn't actually write ended up looking like I did) is due to white space issues. My editor was converting space to tabs, (because I thought we landed on using tabs) which made it look like I changed a lot of lines.

**Comment by prb2** [comment]
This code was merged into repl-inter-datanode and will be reviewed as part of PR #95 

@Swyrough Your comments on this PR were addressed and are a part of that branch. Let me know if you think I missed something.

**Pull request closed without merging**